### ISSUE TITLE
performance question

### ISSUE THREAD

**Issue by drv1234** [f3anaro/pynghttp2#1]
Hi,
first of all thanks for his lib. It makes really easy to use HTTP2 in python.
I hava a http2 framework using h2 (pure python) and now I compared it with this solution.
I used h2load to test an echo server (h2load -n10000 -c3 -m20 http://localhost:8080)
I expected that c binding is at least 10 time faster but it can handle 'only' ~3000msg/sec (the pure h2 soulution can handle 1120).
Using uvloop the result a bit better: 3680 vs 1288.

In h2load page I saw result like:
'finished in 7.08s, 141164.80 req/s, 555.33MB/s' :-O
(However it is possible that the latter was ran on multicore, while the python version is ran on single core.  )

Do you have any measurement or advise how to make it even faster?

thanks

**Comment by f3anaro** [comment]
@drv1234 I very sorry that I answer you one year late :disappointed: Life required my full intention over the last months.

To be honest, I never did performance tests. I used this library primarily as scriptable HTTP/2 client for a study project. Hence, performance was not the main objective.

You can run [cProfile](https://docs.python.org/3/library/profile.html#module-cProfile) and visualize the stats with [KCacheGrind](http://kcachegrind.sourceforge.net/html/Home.html). There is a nice [blog post](https://julien.danjou.info/guide-to-python-profiling-cprofile-concrete-case-carbonara/) explaining the steps in detail. The call graph gives you insights about the bottle necks in the code.

My wild guess is that the Python buffer handling (concatenation of bytes) is the main performance  brake.

**Comment by drv1234** [comment]
hi,
thanks for your reply!

never mind. Unfortunately your solution does not support ssl but it was a requirement for me.
Yes I used cProfile to see the bottleneck.
So meanwhile I changed the code and use nghttp2 python bindings to encode/decode http2 messages (because it is the most time consuming task). And this way I managed to reach 2200 msg/sec (with the whole framework) and this is now enough.

best regards
drv
